core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -663,6 +663,9 @@ void *CudaInternal::resize_team_scratch_space(std::int64_t bytes,
 //----------------------------------------------------------------------------
 
 void CudaInternal::finalize() {
+  // skip if Kokkos::finalize() has already been called
+  if (!singleton().was_finalized) return;
```

Comment:
Does this mean that we are leaking memory in this case (for `m_scratchFlags`, `m_scratchSpace`, `m_scratchConcurrentBitset`, etc.)?

Reply:
Yes

Reply:
This also results in
```
$ ./core/unit_test/KokkosCore_UnitTest_CudaInterOpInit 
========= CUDA-MEMCHECK
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from cuda
[ RUN      ] cuda.raw_cuda_interop
[       OK ] cuda.raw_cuda_interop (278 ms)
[----------] 1 test from cuda (278 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (278 ms total)
[  PASSED  ] 1 test.
Kokkos::Cuda ERROR: Failed to call Kokkos::Cuda::finalize()
```

Reply:
You are right.  Not sure why the CI is passing.

Reply:
I don't see any errors when omitting to skip `finalize` and think that is fine since we are not using the stream for calls to `cudaFree`.

Reply:
Oh I see we just print `std::cerr` and do not actually throw or abort.

Reply:
> I don't see any errors when omitting to skip `finalize` and think that is fine since we are not using the stream for calls to `cudaFree`.

See https://github.com/kokkos/kokkos/pull/3856#discussion_r597967822

I am not sure how to go about this.  I can revert 980deb8 to see if we get that error again.